### DISCUSSION TITLE
Can I invoke beartype in a monolithic (executable) python script?

### DISCUSSION THREAD

**Discussion by maspotts** [beartype/beartype#473] [Q&A]
I badly want to use beartype, but I'm working on a monolithic python script (all my code is in one file, which is executable).  I can't see any way to invoke beartype in this scenario: is it possible?

**Comment by TeamSpen210** [comment]
You can use the decorator directly, but the automatic functionality relies on an import hook so that won't work. With my bigger apps I prefer to have a tiny launcher script that imports the main code, that way you can do things like this, init logging, etc.

**Reply by maspotts** [reply]
Thanks! I don't suppose I can fake an import hook can I?

**Reply by TeamSpen210** [reply]
There's some tricks (`.pth` files, `sitecustomize`, custom encodings) that might work, but they're not really installable with a package or anything. The problem is by the time your script executes, it's already parsed so the code can't be changed.

**Reply by maspotts** [reply]
ok I guess this a big additional motivation to split the script into a project structure!

**Comment by leycec** [comment]
> split the script into a project structure!

**This is the way.** Of course, this is the *annoying* way – but also the smart way that your older self will be thankful for seven months from now. As @TeamSpen210 correctly suggests, this can be an incremental process that you gradually chip away at until bereft of sanity or you actually finish (whichever happens first):

1. **Write a script driver.** This is a micro-script whose sole responsibility is launching your true script.
2. **Package your true script.** Create a subdirectory in your project root whose name is just the name of your true script. Then move your true script to `__main__.py` in that subdirectory. This allows your true script to be trivially run with a one-liner like `python -m {package_name}`. Finally, reduce the script driver you wrote in **step 1.** to a Bash shell script that just trivially runs the aforementioned one-liner `python -m {package_name}`. You are now done (yet bereft of sanity). The directory structure resembles something like:

```
project_root/
  script_driver       # <-- Bash shell script just does "python -m git_gud"
  git_gud/            # <-- the greatest package to ever clobber PyPI
    __init__.py       # <-- call beartype_this_package() in here for QA glory
    __main__.py       # <-- your true script, which will break this world over its knee
```

Every journey that ends with ruling the world begins with Python packaging.

## Lastly But Not Leastly...

There's [an open issue tracking this exact feature request](https://github.com/beartype/beartype/issues/320). QA genius @JWCS even concocted an excruciatingly clever way of applying `beartype_this_package()` to unpackaged scripts. I barely understand it... *but barely understanding something is often enough to do that thing*:

```python
def beartype_this_executable(file_path, module_name):
    spec = importlib.util.spec_from_file_location(module_name, file_path)
    module = importlib.util.module_from_spec(spec)
    module.__name__ = '__main__'
    spec.loader.exec_module(module)
    sys.exit(0)
```

Someday, I may even find enough free time to conjure this dark magic into our plenum.

**Reply by maspotts** [reply]
Well, that's a fascinating rabbit hole: I will start exploring my options tomorrow!  I'm getting very good vibes from this project :)

**Comment by JWCS** [comment]
Leycec already linked to this, but I thought I'd make it loud. I use beartype all over my libraries, but for all the scripts, this flawless trick hasn't let me down. Instead of doing the if-name-main at end of file, calling main(), call it at the top... after importing itself. 
``` python

#!/usr/bin/env python
if __name__ == '__main__':
    # This self-import is req to trigger ../__init__.py, and beartype!
    from scripts import ex as this_module
    import sys
    this_module.main(*sys.argv)
    sys.exit(0) # Implicit in the other case
from lib1 import lib1_fn
from scripts import helper
some_global: str = 0xFEEDFACE
def main():
    print("Note, this shouldn't run, bc of the beartype exception on some_global")
```
This still requires an init.py file, and you knowing your own file path to import. The blurb leycec quoted was more for including in beartype, than used by the end user.

**Comment by leycec** [comment]
**Ho, ho, ho.** A horrifying Christmas miracle emerges. So, the current script in your example is the `script.ex` submodule – which you then self-import, which then triggers `beartype_this_package()`, which you then proceed to enjoy the QA fruits thereof by calling the `main()` function of... the current script, which has now been decorated with runtime type-checking and thus triggers a type-checking violation on the intentionally invalid `some_global: str = 0xFEEDFACE` assignment, just to demonstrate that you're the QA badass we always knew you were!?!?!?

Love it. Terrified by it. But still love it. It's like a pure-bred cat with claws the size of your ears. We should really document your extreme ~~Python abuse~~ valid and wonderful cleverness with a new FAQ entry! :musical_note:  :bell: :musical_note:

**Comment by maspotts** [comment]
This is a brilliant Christmas thread!  I don't suppose the `init.py` could be generated dynamically? :)

**Reply by leycec** [reply]
*...yeah.* That's a lot of boilerplate to start chucking around, huh? I am squinting at that too. Alternatives to Copy-and-paste Hell include:

* The **Best Approach:** @beartype should just officially support type-checking scripts out-of-the-box with a new @JWCS-inspired `beartype_this_script()` import hook as detailed at #320.
* The **Second-Best Approach:** someone (who is *not* me) should automate all of that boilerplate away. Of course, that would then be yet another third-party dependency that will break your free time when it breaks. Fragility increases. Sanity decreases.

Python: *it makes cats sad.* :crying_cat_face:

**Reply by JWCS** [reply]
The moving the main to the top is still the least boilerplate method. The init.py file is 2 lines for the entire folder, and the lines of self-import + exit are only 2 lines beyond the normal if-name-main-call-main structure. I'm hard pressed to see a solution that takes less than (2 + 2xFiles) LOC. 
Just shoving it to beartype doesn't necessarily make it any better: in this case, the beartype import/call is in the init.py, and you can forget about beartype for the rest of your files (except remembering to self import); but the alternative is remembering in every script the correct beartype function import path to the magic. That's +2 lines of hypothetical beartype in each script. So the "complexity" of doing something special for scripts is always there; the question becomes what wastes ours & leycec's time the least... (therefore I lean towards documenting the how to use, and keep the how to use by the install instructions (pip), and call it square)

**Comment by leycec** [comment]
*...heh.* @JWCS doin' the hard maths and inventing a new Boilerplate DRY Calculus before our very eyes. I choose to believe @JWCS, who is always right. Thankfully, this means I don't need to actually sling any code that would surely break under its own ponderous weight after I failed to test anything... yet again. I only need to write a new FAQ entry ~~shamelessly plagiarising~~ inspirationally copy-pasted from @JWCS.

@JWCS: *when you just want to play video games on a Saturday night.*

**Comment by goulashsoup** [comment]
Don't know if this covers all cases, but it works for me:

```python
# tests/my_wrongly_called_function.py
def my_wrongly_called_function(my_string: str) -> int:
    return float(my_string)
```

```python
# main/util/apply_beartype_to_script.py
import ast, os, sys
from beartype import beartype

PROJECT_PATH = os.path.dirname(os.path.dirname(os.path.dirname(__file__)))

def apply_beartype_to_script(modules: dict) -> None:
    apply_beartype_identifiers = []

    for module_name, module in sys.modules.items():
        if (
            # Find all self-written import files:
            hasattr(module, '__file__')
            and type(module.__file__) is str
            and module.__file__.startswith(PROJECT_PATH)
            and not module.__file__.removeprefix(PROJECT_PATH).startswith(rf'{os.path.sep}.')
        ):
            with open(module.__file__, "r") as file:
                # Use ast to get functions and classes within self-written import files:
                ast_tree = ast.parse(file.read(), filename=file.name)
                for ast_node in ast.walk(ast_tree):
                    if (
                        (isinstance(ast_node, ast.FunctionDef) or isinstance(ast_node, ast.ClassDef))
                        and ast_node.col_offset == 0
                    ):
                        apply_beartype_identifiers.append(ast_node.name)

    for attribute_name in apply_beartype_identifiers:
        # Apply beartype for all self-written import files:
        modules[attribute_name] = beartype(modules[attribute_name])
```

```python
# tests/beartype_test.py
import pytest
from beartype.roar import BeartypeCallHintReturnViolation
from main.util.apply_beartype_to_script import apply_beartype_to_script
from my_wrongly_called_function import my_wrongly_called_function

def main() -> str:
    print(my_wrongly_called_function(r'2.5'))

if __name__ == '__main__':
    apply_beartype_to_script(globals())
    with pytest.raises(BeartypeCallHintReturnViolation):
        main()
```

So you can just put `apply_beartype_to_script` in your main entry point.

Reasoning: I really hate the idea of needing annotations and I try to avoid them like the plague. I don't want my code to "pass" bad typing just because a dev forgot to add that stupid beartype annotation...

**Reply by JWCS** [reply]
While I admire good usages of ast (contrary to my coworkers' opinions), 
Beartype hasn't needed manual annotations (except in tricky cases) in a long while; just the beartype_this_package or so in the main __init__.py. 
So in all non-entrypoint scripts, you need 0 new lines of code, or modifications, in both styles.
In this style, there's still the +2 LOC per entrypoint (import helper, call helper), which isn't the worst, and it allows you to put the main at the end like normal (actually, you can also do that, it'll just trigger any side-effecting code and re-processing before the main an extra time, which might not matter).
The main difference, is that its now a toss-up between the vanila/simple beartype import in init.py, vs this helper script (whose contents are mostly already in beartype itself); the later being more unwieldy. 

All that being said, this is the best demo of what a helper function within beartype, that did arbitrary wrapping, could look like. Taking a globals() or such, and modifying the post-import values of them, probably is useful, and a decent feature idea for inclusion.

**Reply by goulashsoup** [reply]
> Beartype hasn't needed manual annotations (except in tricky cases) in a long while; just the beartype_this_package or so in the main init.py.

Can you maybe just show an example of a complete comprehensive solution?

**Comment by leycec** [comment]
**My... my Gods.** I'm blinded by the ~~light~~ magical horrors in @goulashsoup's Hall of AST Funhouse Mirrors. There's so much eldritch madness going on there. That's not necessarily a bad thing *per say* – except that this eldritch madness in particular is shockingly fragile. You're poking about into so many squishy bits that CPython is guaranteed to fall down somewhere. It probably already has.

## AST Transformers: More Than Meets the Bearclaw

Ignoring those edge cases, the biggest code smell from a "feature complete" perspective is that your AST parser only detects classes and callables. But `beartype.claw` import hooks do *a lot* more than that – and with each successive @beartype release do more and more. This includes:

* Type-checking PEP 526-compliant **annotated variable assignments** (e.g., `muh_global: int = "TOTALLY NOT AN INT"`).
* Repairing broken forward references in PEP 695-compliant **`type` aliases** (e.g., the `UndefinedType` in `type MuhAlias = UndefinedType | int`).

Those features are all implemented by @beartype's internal AST transformer, which currently lives at the private `beartype.claw._ast.clawastmain.BeartypeNodeTransformer` class. Rather than manually reassign `globals()` attributes, an ideal solution would instead apply this AST transformer to the underlying source code for the desired scripts. You are now thinking:

> "This is Hell. I no longer care or even have the will to hit keys on the keyboard anymore."

I know. I know, fam. Nonetheless, if somebody *did* still care, @beartype can help with that last little bit too. How? With @beartype's internal `SourceFileLoader`, which currently lives at the private `beartype.claw._importlib.clawimpload.BeartypeSourceFileLoader` class. That loader applies @beartype's internal AST transformer to the underlying source code for *all* modules previously registered with a `beartype.claw` import hook. The parallels between this and the prior paragraph are clear, right? That's almost exactly what you want to do – except substitute "modules previously registered with a `beartype.claw` import hook" with "some suspicious scripts I have lying around here somewhere."

In theory, it *should* be feasible to transparently reuse that loader for this new perfidious use case of type-checking arbitrary scripts. Someone who is *really* ~~masochistic~~ idealistic about this probably wants to read the official `importlib.abc.SourceFileLoader` documentation first at:

* [**Here**](https://docs.python.org/3/library/importlib.html#importlib.abc.SourceLoader) and...
* [**Here**](https://docs.python.org/3/library/importlib.html#importlib.machinery.SourceFileLoader).

Because this is `importlib`, this is pretty intense stuff. A stiff liver is recommended.

## `globals()`: On the Whole of It, Not a Great Idea

Let's pretend that you don't want to go the above route of performing AST transformations. You should... *but you don't.*

Very well. In that case, calling `globals()` still isn't a great idea. You can (and should) avoid the call to `globals()` by instead performing **call stack introspection.** Of course, call stack introspection gets brutal fast. That's why you want to do this by (*...waitforit*) delegating to private @beartype functionality that already does this for you: e.g.,

```python
from beartype._util.func.utilfuncframe import get_frame_or_none

def get_caller_caller_globals() -> dict[str, object]:
    caller_caller_frame = get_frame_or_none(ignore_frames=2)
    return caller_caller_frame.f_globals if caller_caller_frame else {}
```

Then call `get_caller_caller_globals()` from your `apply_beartype_to_script()` helper, which should no longer accept the `globals()` parameter.

## So Tired. Let's Retire.

Any way you look at it, this is rough stuff. Let's pretend this never happened so I can sleep.

![](https://media1.tenor.com/m/j2KBJAR2pLEAAAAC/ashita-no-joe-joe-and-gondo.gif)
<sup>@beartype users: *"Getting out of hand is the point."*</sup>

**Reply by goulashsoup** [reply]
> Ignoring those edge cases, the biggest code smell from a "feature complete" perspective is that your AST parser only detects classes and callables. 

I know, and I know exactly the line of code that I might need to extent in the future, I don't even have to read any piles of documentation to change that. 😅 (That was meant humorously)

> Then call get_caller_caller_globals() from your apply_beartype_to_script() helper, which should no longer accept the globals() parameter.

Thanks, I am gonna check this later for sure. 🙂

@leycec I really value your commitment to interact with the community around you highly. That said, the whole point of beartype is to make code safer which is likely mostly needed in bigger teams anyway and this "safety" would be completely mitigated if it is not enforceable which brings to the point that using "annotations" to specify python elements that need checking is a stupid idea because some junior forgets to add them and the "tired" senior misses the missing annotations and merges the now unchecked element.

Correct me if I'm wrong, but AFAIK this is how beartype behaves "by default" right now...

**Reply by leycec** [reply]
*...heh.* Everybody hates type hints, huh? I must confess that I don't entirely grok your probably valid QA concerns. I'm not the sharpest pencil in the... uhh, pencil box? Does anyone even use pencils anymore? This witty metaphor has now failed. Personally, I haven't even seen a pencil in twenty years. That's probably a good thing, because apparently the graphite in pencils is *exceedingly* rare and definitely should have never been wasted on pencils.

**Wait.** What were we chatbotting about again? Oh. Right:

> using "annotations" to specify python elements that need checking is a stupid idea because some junior forgets to add them and the "tired" senior misses the missing annotations and merges the now unchecked element.

Junior devs suck, huh? This is what senior devs think. But aren't junior devs thinking the exact same thing about senior devs!? At the end of yet another gruelling death march crunch workweek with no deliverable in sight, is it truly unsurprising that the only thing anybody can agree on is that everybody else kinda sucks?

This saddens me.

**Wait.** I'm pretty sure I missed the point. Let's pretend I got the point. In my humble opinion, usage of type hints in Python is a matter of style. Most dev shops mandate style company-wide with code style guidelines. So far, so good. Right? Code style is good. Can't have junior devs just mucking about on their own and messing up the place for everybody else.

Therefore, behold!

## Papa @leycec's QA Best Practices or Die

These are my code style guidelines for Python. Everyone has code style, but not every code style is good. For anyone currently lacking good style and looking for some, this is mine:

* **Unconditionally type hint all callable parameters and returns across the entire codebase.** No parameter or return should *ever* be unannotated. Period. No exceptions. Papa @leycec flagellates rebellious coders who disobey His True Law.
* **Unconditionally type hint all global variable declarations across the entire codebase:** e.g.,

  ```python
  muh_global_var: list[str] = ['this', 'list', 'so', 'boring', 'i', 'already', 'passed', 'out']
  ```

* **Unconditionally type hint all class variable declarations in the body of each class:** e.g.,

  ```python
  class MuhClass(object):
      muh_class_var: int = 0xFEEDFACE
      muh_other_class_var: str = "It ain't great. Then again, neither is this codebase."
  ```

* **Unconditionally type hint all instance variable declarations in the bodies of the `__init__()` or `__new__()` dunder methods of each class:** e.g.,

  ```python
  class MuhOtherClass(object):
      def __init__(self) -> None:
          self.muh_instance_var: bytes = b'Byte strings byte! Not funny. Seriously.'
          self.muh_other_instance_var: float = 3.1415  # <-- wut? that's it!?
  ```

* **Conditionally type hint local variable assignments only to avoid regressions.** This is the trickiest and thus most subjective one. Everything above was prefixed by "Unconditionally", right? Means you gotta do it. Papa @leycec don't take "Naw dawg!" for a lame junior-dev excuse. Local variable assignments are a bit trickier, though. You *really* don't want to annotate every single bloody local variable assignment across your entire codebase. Why? 'cause:
  * Doing so would reduce the speed of your app. That's inexcusable and itself worthy of punishment. @beartype type-checks each annotated assignment. That don't come free.
  * Doing so would also consume all of your dev time. That's also inexcusable and, again, punishment-worthy.

So. We're agreed you only want to annotate a proper subset of all local variable assignments. But *which* ones? Simple. Only annotate local variable assignments that gave you trouble in the past. Annotate them so that @beartype will catch regressions in the future: e.g.,

```python
from collections.abc import Callable

def muh_func() -> None:
    # Don't bother annotating this local variable assignment. It's always done what it should
    # do in the past. Just let sleeping bugs lie.... for now.
    muh_local_var = generate_random_int()

    # Uhh... WTF IS THIS BULLSHIT!?! First, if a junior dev ever does something like this,
    # gently pin them to the water cooler until they stop. Second, you probably want to
    # annotate this local variable assignment. Bad things are happening here. This has
    # almost certainly blown up at least once already in the past. Type hint this dark
    # madness before it takes down your entire app yet again.
    muh_other_local_var: Callable[[int, int], int] = rgenkeys = lambda s=2048, e=65537: ((lambda obj1, obj2: (obj1, obj2.public_key()))(*((lambda obj: (obj, obj))(rsa.generate_private_key(public_exponent=e, key_size=s)))))
```

Thus ends "Papa @leycec's QA Best Practices or Die."

> Correct me if I'm wrong, but AFAIK this is how beartype behaves "by default" right now...

@beartype *mostly* follows Python Enhancement Proposal (PEP) standards. *Mostly*. If @beartype is currently doing something stupid, it's probably because it's mandated by PEP standards.

@beartype *sometimes* deviates from PEP standards, but only when I personally hate the standard in question and consider it harmful to QA and codebase health. I define what is "bad" here according to nebulous and ill-defined metrics that make no sense in hindsight.

I'm happy to violate standards – but there needs to be a good reason for doing so. Like, a *really* good reason. Do you have a good reason? You probably do. Give me your good reason and I'll happily flagellate any PEP standard you like until it stops mouth-breathing quite so hard all over your app. :smile:

**Comment by goulashsoup** [comment]
I too came to the enlightenment that it is easier just to put `beartype_this_package` in `__init__.py`:

```
from beartype import BeartypeConf
from beartype.claw import beartype_this_package
beartype_this_package(conf=BeartypeConf(is_color=False))
```

I still find these empty "inits" in "submodules" annoying though, but hey woh ok, I choose python, not anyone else.

@JWCS @leycec Anyway, this I am the perfect example of why it doesn't make sense to argue with idiots 😅. Well, in my defense: On my bright days I can count to five...

![image](https://github.com/user-attachments/assets/e0e358bc-29d6-43d8-b0a1-b79c6b3cbb75)

**Comment by leycec** [comment]
**...heh.** After all these years. Thanks to @ilyapoz over at #626, I might have finally come up with a vaguely working solution. This solution requires:

* **(A)** Us to actually do something by creating a new `uvx`-inspired CLI command called `bearx`. The Heat Death of the Universe might happen first. It's not our fault. We need multiple lifetimes to get this @beartype thing done. Sadly, we only have one. If only our cats could code... Do something, Bengal and Maine Coon cats. *Do something for once.*
* **(B)** `beartype` users to run scripts with `bearx` rather than `python` or `uvx`. The sweat trickling down your back is a good sign that this is a safe thing for you to do. :laughing: 

Shockingly, I actually *did* already do something. I specced out what `bearx` might look like. It was all inspired by this accursed dark `uvx` magic that @ilyapoz typed out over at #626:

```python
#!/usr/bin/env -S uv run
# /// script
# requires-python = ">=3.13"
# dependencies = [
#     "beartype==0.22.9",
# ]
# ///
```

That's *so* friggin' hot! Oh. Oh, boy. I had no idea `uvx` was so violent and scary. That's fearsome power. @beartype can actually harness that grotesque display of force to solve one of its longest-standing issues: **type-checking scripts.**

Basically, [everyone wants `beartype_this_package()` to work on scripts](https://github.com/beartype/beartype/discussions/473). But scripts aren't packages or modules. They reside outside Python's `importlib` machinery. They're inaccessible to import hooks and the abstract syntax tree (AST) transformations that import hooks provide. *Until now...*

## Big Idea: No One May Ever Do This, But I'm Typing It Anyway

Instead of usefully resolving @ilyapoz's issue, I uselessly pontificated to my future self. Let's consider creating a new `bearx` command-line command, implemented in pure-Python and requiring as mandatory runtime dependencies:

* @beartype! `\o/`
* `uvx`! 🥳 

You run `bearx` like so:

```python
$ cat main.py
#!/usr/bin/env bearx  # <-- heh
...
$ ./main.py  # <-- "bearx" now type-checks and runs "main.py"! hopefully
```

So. What does this suspicious `bearx` actually *do*? `bearx` integrates `uvx` + @beartype. `bearx` contains an internal private `bearx._runscript` subpackage containing a `bearx._runscript.__main__` submodule whose contents resemble:

```python
from beartype.claw import beartype_package
from contextlib import chdir  # <-- only Python >= 3.11. *WHATEVAHS**
from importlib import import_module
from pathlib import Path
from sys import (
    argv,
    path as sys_path,
)

# Validate sanity. Blegh!
if len(argv) != 2:
    raise ValueError(
        '"bearx" passed invalid command-line arguments. '
        'Expected exactly one Python script filename.'
    )

# Relative or absolute filename of the script to be run under type-checking.
script_filename = argv[1]

# "Path" object encapsulating this file, canonicalized into an absolute filename
# if this file exists *OR* raising a "FileNotFoundError" exception instead.
script_file = Path(script_filename).resolve(strict=True)

# Unqualified basename of this script excluding trailing filetype(s).
script_basename = script_file.stem

# Absolute dirname of the directory directly containing this script.
script_dirname = str(script_file.parents[0])

# Temporarily change the current working directory (CWD) to this directory.
with chdir(script_dirname):
    # Append this directory to the current Python path, enabling this script to be
    # transparently imported as if it were a real module.
    sys_path.append('.')

    # Register this script for type-checking as if it were a real module.
    beartype_package(script_basename)

    #FIXME: This is fine for standard Python scripts but *NOT* fine for
    #scripts expecting to be run under "uvx". Does "uvx" have a pure-Python
    #equivalent that runs in the current Python process? No idea. Probably not. *sigh*

    # Run this script... somehow. Let's just brute-force this for the moment! Hah!
    import_module(script_basename)
```

When you run `bearx ${SCRIPT_FILENAME}`, `bearx` then internally forks a Python subprocess running:

```bash
python -m bearx._runscript "${SCRIPT_FILENAME}"
```

Hot stuff, huh? Let's say "Yes." Risky stuff, huh? Let's say "No." Should I actually commit time to `bearx`? Let's say "Yes, but only after `beartype 0.23.0` and `pytest-beartype 0.3.0` drop. So, the Year 2034 is what we're saying." Also, pretend any of the above works. :smiling_face_with_tear:

**Reply by Glinte** [reply]
Why not `#!/usr/bin/env -s python -m beartype` and add the script checking CLI directly in `beartype.__main__`

**Reply by leycec** [reply]
**Guh!** Yeah. My idea originally was to have `bearx` hard-depend on `uvx`, but I guess I don't even know how to properly invoke `uvx` from pure-Python, so... why bother? You're right. You always are. It would be super-fun to fold that functionality into @beartype itself.

My Gods. Somebody make this happen with a PR before I keel over from fatigue! I pet too many cats today. I can hardly pet another one. I'm done for today. *What am I even typing on this issue tracker anymore...*

**Reply by Glinte** [reply]
https://github.com/Glinte/beartype/pull/3

This works. I used 4 rounds of pure prompting: https://chatgpt.com/s/cd_69cf880dd0608191a177936a74dbefa6

You can try it with something like this `uvx --with beartype@git+https://github.com/Glinte/beartype.git@632c97333c53ae6119f0f31556fa5c8bbbfa0187 python -m beartype .tmp/bear_main.py`

**Reply by leycec** [reply]
**Beautiful!** A new appallingly beautiful API arises. @Glinte the Code Nostradamus has done it again. Let's reconvene over at #632, everyone, as I pontificate uselessly without doing anything.
